File: app/src/main/java/com/kickstarter/viewmodels/ProjectNotificationViewModel.java
package com.kickstarter.viewmodels;

import com.kickstarter.libs.ActivityViewModel;
import com.kickstarter.libs.Environment;
import com.kickstarter.models.ProjectNotification;
import com.kickstarter.services.ApiClientType;
import com.kickstarter.ui.viewholders.ProjectNotificationViewHolder;

import androidx.annotation.NonNull;
import rx.Notification;
import rx.Observable;
import rx.subjects.BehaviorSubject;
import rx.subjects.PublishSubject;

import static com.kickstarter.libs.rx.transformers.Transformers.errors;
import static com.kickstarter.libs.rx.transformers.Transformers.takePairWhen;
import static com.kickstarter.libs.rx.transformers.Transformers.values;

public interface ProjectNotificationViewModel {
  interface Inputs {
    /** Call when the enable switch is clicked. */
    void enabledSwitchClick(boolean enabled);

    /** Call when a notification is bound to the viewholder. */
    void projectNotification(ProjectNotification projectNotification);
  }

  interface Outputs {
    /** Emits `True` if the enabled switch should be toggled on, `False` otherwise. */
    Observable<Boolean> enabledSwitch();

    /** Emits the project's name. */
    Observable<String> projectName();

    /**  Show an error indicating the notification cannot be saved. */
    Observable<Void> showUnableToSaveProjectNotificationError();
  }

  final class ViewModel extends ActivityViewModel<ProjectNotificationViewHolder> implements Inputs, Outputs {
    public ViewModel(final @NonNull Environment environment) {
      super(environment);

      final ApiClientType client = environment.apiClient();

      // When the enable switch is clicked, update the project notification.
      final Observable<Notification<ProjectNotification>> updateNotification = this.projectNotification
        .compose(takePairWhen(this.enabledSwitchClick))
        .switchMap(ne ->
          client
            .updateProjectNotifications(ne.first, ne.second)
            .materialize()
        )
        .share();

      updateNotification
        .compose(values())
        .compose(bindToLifecycle())
        .subscribe(this.projectNotification::onNext);

      updateNotification
        .compose(errors())
        .compose(bindToLifecycle())
        .subscribe(__ -> this.showUnableToSaveProjectNotificationError.onNext(null));

      // Update the project name when a project notification emits.
      this.projectNotification
        .map(n -> n.project().name())
        .compose(bindToLifecycle())
        .subscribe(this.projectName::onNext);

      // Update the enabled switch when a project notification emits.
      this.projectNotification
        .map(n -> n.email() && n.mobile())
        .compose(bindToLifecycle())
        .subscribe(this.enabledSwitch::onNext);
    }

    private PublishSubject<Boolean> enabledSwitchClick = PublishSubject.create();
    private PublishSubject<ProjectNotification> projectNotification = PublishSubject.create();

    private BehaviorSubject<String> projectName = BehaviorSubject.create();
    private BehaviorSubject<Boolean> enabledSwitch = BehaviorSubject.create();

    private PublishSubject<Void> showUnableToSaveProjectNotificationError = PublishSubject.create();

    public Inputs inputs = this;
    public Outputs outputs = this;

    @Override public void enabledSwitchClick(final boolean enabled) {
      this.enabledSwitchClick.onNext(enabled);
    }
    @Override public void projectNotification(final @NonNull ProjectNotification projectNotification) {
      this.projectNotification.onNext(projectNotification);
    }

    @Override public @NonNull Observable<String> projectName() {
      return this.projectName;
    }
    @Override public @NonNull Observable<Boolean> enabledSwitch() {
      return this.enabledSwitch;
    }
    @Override public @NonNull Observable<Void> showUnableToSaveProjectNotificationError() {
      return this.showUnableToSaveProjectNotificationError;
    }
  }
}

File: app/src/main/java/com/kickstarter/viewmodels/ProjectNotificationViewModel.kt
package com.kickstarter.viewmodels

import com.kickstarter.libs.Environment
import com.kickstarter.libs.rx.transformers.Transformers
import com.kickstarter.libs.rx.transformers.Transformers.errorsV2
import com.kickstarter.libs.rx.transformers.Transformers.valuesV2
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.isNotNull
import com.kickstarter.models.ProjectNotification
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.PublishSubject

interface ProjectNotificationViewModel {
    interface Inputs {
        /** Call when the enable switch is clicked.  */
        fun enabledSwitchClick(enabled: Boolean)

        /** Call when a notification is bound to the viewholder.  */
        fun projectNotification(projectNotification: ProjectNotification)
    }

    interface Outputs {
        /** Emits `True` if the enabled switch should be toggled on, `False` otherwise.  */
        fun enabledSwitch(): Observable<Boolean>

        /** Emits the project's name.  */
        fun projectName(): Observable<String>

        /**  Show an error indicating the notification cannot be saved.  */
        fun showUnableToSaveProjectNotificationError(): Observable<Unit>
    }

    class ViewModel(environment: Environment) : androidx.lifecycle.ViewModel(), Inputs, Outputs {

        private val enabledSwitchClick = PublishSubject.create<Boolean>()
        private val projectNotification = PublishSubject.create<ProjectNotification>()
        private val projectName = BehaviorSubject.create<String>()
        private val enabledSwitch = BehaviorSubject.create<Boolean>()
        private val showUnableToSaveProjectNotificationError = PublishSubject.create<Unit>()
        private val client = requireNotNull(environment.apiClientV2())
        private val disposables = CompositeDisposable()

        var inputs: Inputs = this
        var outputs: Outputs = this

        init {
            // When the enable switch is clicked, update the project notification.
            val updateNotification = projectNotification
                .compose(Transformers.takePairWhenV2(enabledSwitchClick))
                .switchMap {
                    client.updateProjectNotifications(it.first, it.second)
                }
                .materialize()
                .share()

            updateNotification
                .compose(valuesV2())
                .filter { it.isNotNull() }
                .subscribe { projectNotification.onNext(it) }
                .addToDisposable(disposables)

            updateNotification
                .compose(errorsV2())
                .subscribe { showUnableToSaveProjectNotificationError.onNext(Unit) }
                .addToDisposable(disposables)

            // Update the project name when a project notification emits.
            projectNotification
                .filter { it.project().isNotNull() && it.project().name().isNotNull() }
                .map { it.project().name() }
                .subscribe { projectName.onNext(it) }
                .addToDisposable(disposables)

            // Update the enabled switch when a project notification emits.
            projectNotification
                .map { it.email() && it.mobile() }
                .subscribe { enabledSwitch.onNext(it) }
                .addToDisposable(disposables)
        }

        override fun enabledSwitchClick(enabled: Boolean) {
            enabledSwitchClick.onNext(enabled)
        }

        override fun projectNotification(projectNotification: ProjectNotification) {
            this.projectNotification.onNext(projectNotification)
        }

        override fun projectName(): Observable<String> = projectName

        override fun enabledSwitch(): Observable<Boolean> = enabledSwitch

        override fun showUnableToSaveProjectNotificationError(): Observable<Unit> = showUnableToSaveProjectNotificationError

        override fun onCleared() {
            super.onCleared()
            disposables.clear()
        }
    }
}
